cpp/cmake_modules/SetupCxxFlags.cmake

```diff
@@ -118,12 +118,16 @@ if(NOT DEFINED CMAKE_C_STANDARD)
   set(CMAKE_C_STANDARD 11)
 endif()
 
-# This ensures that things like c++11 get passed correctly
+# This ensures that things like c++11/c++14 get passed correctly
 if(NOT DEFINED CMAKE_CXX_STANDARD)
-  set(CMAKE_CXX_STANDARD 11)
+  if(ARROW_AZURE)
+    set(CMAKE_CXX_STANDARD 14)
```

Comment:
Hmm... I don't think we should do that over the entire codebase, only `azurefs.cc`?

Reply:
Is it possible to set `CMAKE_CXX_STANDARD` to 14 only for the Azure component given that `azurefs.cc`/`azurefs_test.cc` is a part of the Arrow target? If we make a separate target for compiling Azure files then it might be possible to use `set_target_properties(azurefs_objlib PROPERTIES CXX_STANDARD 14 CXX_STANDARD_REQUIRED ON)`. Initially, we had a separate Azure target but this change was suggested by @kou

Reply:
I have no idea what CMake allows but hopefully we are not obliged to make this a project-wide decision. @kou What is your take?

Reply:
Or we can require CMake 3.12 or later for `-DARROW_AZURE=ON` and use object library instead of project global `CMAKE_CXX_STANDARD=14`.

Reply:
That might be better. Upgrading CMake is easier than changing compilers.

Reply:
Note that we need a C++14 compatible compiler for `-DARROW_AZURE=ON` even if we choose either the `CMAKE_CXX_STANDARD=14` approach or the object library approach.

Reply:
So we may as well keep it like that then.